# Ensemble Learning

## Initial Imports

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [3]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

## Read the CSV and Perform Basic Data Cleaning

In [25]:
# Load the data
pd.set_option('max_columns', None)
file_path = Path('Resources/LoanStats_2019Q1.csv')
df = pd.read_csv(file_path)


loan_amnt  int_rate  installment home_ownership  annual_inc  \
23702    20000.0    0.0819       628.49           RENT     98000.0   
46065     6600.0    0.1180       218.59           RENT     35000.0   
65412    13000.0    0.1298       295.66       MORTGAGE    132000.0   
41498    20000.0    0.0702       617.73       MORTGAGE    135000.0   
57017     6500.0    0.1691       231.46       MORTGAGE     34776.0   
46779    18500.0    0.0819       581.35           RENT     45000.0   
4590      8000.0    0.2250       307.60       MORTGAGE    105000.0   
67612    18000.0    0.1447       423.23       MORTGAGE     75000.0   
54328     9000.0    0.1180       298.07       MORTGAGE    159000.0   
50305     6500.0    0.1614       228.98       MORTGAGE    100000.0   

      verification_status   issue_d loan_status pymnt_plan    dti  \
23702        Not Verified  Feb-2019    low_risk          n  13.15   
46065     Source Verified  Jan-2019    low_risk          n  20.51   
65412            Verified  Jan-2019    low_risk          n   7.15   
41498            Verified  Jan-2019    low_risk          n  19.21   
57017        Not Verified  Jan-2019    low_risk          n  20.67   
46779        Not Verified  Jan-2019    low_risk          n  24.69   
4590      Source Verified  Mar-2019    low_risk          n  10.13   
67612        Not Verified  Jan-2019    low_risk          n  13.71   
54328        Not Verified  Jan-2019    low_risk          n  20.33   
50305     Source Verified  Jan-2019    low_risk          n  39.30   

       delinq_2yrs  inq_last_6mths  open_acc  pub_rec  revol_bal  total_acc  \
23702          0.0             2.0      18.0      1.0    19676.0       27.0   
46065          1.0             0.0      10.0      0.0     4496.0       25.0   
65412          0.0             1.0      18.0      0.0    11180.0       33.0   
41498          0.0             1.0      19.0      0.0   175165.0       32.0   
57017          0.0             0.0       8.0      1.0     7820.0       15.0   
46779          0.0             1.0       7.0      0.0     9220.0       13.0   
4590           0.0             1.0       8.0      0.0    10078.0       14.0   
67612          1.0             0.0      13.0      0.0    15956.0       20.0   
54328          0.0             0.0       5.0      0.0    26127.0       34.0   
50305          0.0             1.0      24.0      0.0    60506.0       42.0   

      initial_list_status  out_prncp  out_prncp_inv  total_pymnt  \
23702                   w   19508.01       19508.01       619.39   
46065                   w    6291.11        6291.11       432.85   
65412                   w   12529.82       12529.82       868.23   
41498                   w   18995.61       18995.61      1258.86   
57017                   w    6074.53        6074.53       688.27   
46779                   w   17586.71       17586.71      1183.75   
4590                    w    7827.40        7827.40       297.60   
67612                   w   17373.97       17373.97      1349.28   
54328                   w    8365.09        8365.09       888.31   
50305                   w    6041.73        6041.73       681.11   

       total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
23702           619.39           491.99         127.40                 0.0   
46065           432.85           308.89         123.96                 0.0   
65412           868.23           470.18         398.05                 0.0   
41498          1258.86          1004.39         254.47                 0.0   
57017           688.27           425.47         262.80                 0.0   
46779          1183.75           913.29         270.46                 0.0   
4590            297.60           172.60         125.00                 0.0   
67612          1349.28           626.03         723.25                 0.0   
54328           888.31           634.91         253.40                 0.0   
50305           681.11           458.27         222.84                 0.0   

In [32]:
# Preview the data
df.sample(10)

loan_amnt  int_rate  installment home_ownership  annual_inc  \
39646     4200.0    0.1131       138.13       MORTGAGE     33258.0   
22733     7500.0    0.0819       235.69       MORTGAGE     61000.0   
10846    16000.0    0.1557       385.45       MORTGAGE    125000.0   
22623     6250.0    0.1797       225.86           RENT     80000.0   
64723    17625.0    0.2235       490.30       MORTGAGE     77000.0   
49105    28000.0    0.0819       570.29            OWN     80000.0   
44994     3300.0    0.1691       117.51       MORTGAGE     52000.0   
61912    12000.0    0.0702       370.64       MORTGAGE     10000.0   
48638    15000.0    0.1691       534.12           RENT     35000.0   
68522    22000.0    0.0819       691.33           RENT     55000.0   

      verification_status   issue_d loan_status pymnt_plan     dti  \
39646        Not Verified  Feb-2019    low_risk          n   26.71   
22733     Source Verified  Feb-2019    low_risk          n   30.26   
10846     Source Verified  Mar-2019    low_risk          n   19.42   
22623        Not Verified  Feb-2019    low_risk          n    9.62   
64723        Not Verified  Jan-2019    low_risk          n   35.10   
49105        Not Verified  Jan-2019    low_risk          n   22.71   
44994     Source Verified  Jan-2019    low_risk          n   30.83   
61912            Verified  Jan-2019    low_risk          n  396.52   
48638     Source Verified  Jan-2019    low_risk          n    1.41   
68522        Not Verified  Jan-2019    low_risk          n   14.82   

       delinq_2yrs  inq_last_6mths  open_acc  pub_rec  revol_bal  total_acc  \
39646          0.0             0.0      11.0      0.0     4731.0       16.0   
22733          0.0             0.0      13.0      0.0    13418.0       17.0   
10846          0.0             2.0      27.0      1.0    16637.0       83.0   
22623          0.0             1.0       9.0      0.0     5707.0       14.0   
64723          0.0             0.0      26.0      0.0    44689.0       36.0   
49105          0.0             0.0      10.0      0.0    33345.0       26.0   
44994          0.0             1.0       9.0      0.0     6083.0       19.0   
61912          0.0             0.0      25.0      0.0    82587.0       36.0   
48638          0.0             4.0       3.0      0.0     4110.0       33.0   
68522          0.0             0.0       5.0      0.0     4972.0       17.0   

      initial_list_status  out_prncp  out_prncp_inv  total_pymnt  \
39646                   f    4001.98        4001.98       270.98   
22733                   w    7315.50        7315.50       232.28   
10846                   w   15822.15       15822.15       371.61   
22623                   w    6117.73        6117.73       219.62   
64723                   w   17215.70       17215.70      1449.02   
49105                   w   27239.03       27239.03      1127.84   
44994                   w    3156.98        3156.98       228.82   
61912                   f   11093.40       11093.40      1107.24   
48638                   f   14349.96       14349.96      1040.06   
68522                   w   19432.51       19432.51      2989.99   

       total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
39646           270.98           198.02          72.96                 0.0   
22733           232.28           184.50          47.78                 0.0   
10846           371.61           177.85         193.76                 0.0   
22623           219.62           132.27          87.35                 0.0   
64723          1449.02           409.30        1039.72                 0.0   
49105          1127.84           760.97         366.87                 0.0   
44994           228.82           143.02          85.80                 0.0   
61912          1107.24           906.60         200.64                 0.0   
48638          1040.06           650.04         390.02                 0.0   
68522          2989.99          2567.49         422.50            

In [30]:
columns = [
    "loan_amnt", "int_rate", "installment", "home_ownership",
    "annual_inc", "verification_status", "issue_d", "loan_status",
    "pymnt_plan", "dti", "delinq_2yrs", "inq_last_6mths",
    "open_acc", "pub_rec", "revol_bal", "total_acc",
    "initial_list_status", "out_prncp", "out_prncp_inv", "total_pymnt",
    "total_pymnt_inv", "total_rec_prncp", "total_rec_int", "total_rec_late_fee",
    "recoveries", "collection_recovery_fee", "last_pymnt_amnt", "next_pymnt_d",
    "collections_12_mths_ex_med", "policy_code", "application_type", "acc_now_delinq",
    "tot_coll_amt", "tot_cur_bal", "open_acc_6m", "open_act_il",
    "open_il_12m", "open_il_24m", "mths_since_rcnt_il", "total_bal_il",
    "il_util", "open_rv_12m", "open_rv_24m", "max_bal_bc",
    "all_util", "total_rev_hi_lim", "inq_fi", "total_cu_tl",
    "inq_last_12m", "acc_open_past_24mths", "avg_cur_bal", "bc_open_to_buy",
    "bc_util", "chargeoff_within_12_mths", "delinq_amnt", "mo_sin_old_il_acct",
    "mo_sin_old_rev_tl_op", "mo_sin_rcnt_rev_tl_op", "mo_sin_rcnt_tl", "mort_acc",
    "mths_since_recent_bc", "mths_since_recent_inq", "num_accts_ever_120_pd", "num_actv_bc_tl",
    "num_actv_rev_tl", "num_bc_sats", "num_bc_tl", "num_il_tl",
    "num_op_rev_tl", "num_rev_accts", "num_rev_tl_bal_gt_0",
    "num_sats", "num_tl_120dpd_2m", "num_tl_30dpd", "num_tl_90g_dpd_24m",
    "num_tl_op_past_12m", "pct_tl_nvr_dlq", "percent_bc_gt_75", "pub_rec_bankruptcies",
    "tax_liens", "tot_hi_cred_lim", "total_bal_ex_mort", "total_bc_limit",
    "total_il_high_credit_limit", "hardship_flag", "debt_settlement_flag"
]

target = ["loan_status"]

In [6]:
 df.shape

(68817, 86)

In [7]:
 df = df.dropna(axis='columns', how='all')

In [8]:
df.shape

(68817, 86)

In [9]:
df = df.dropna()

In [10]:
df.shape

(68817, 86)

## Split the Data into Training and Testing

In [33]:
# Create our features
# X_scaler wasn't working because home_ownership had 3 values - couldn't go binary
# X = df.drop(columns='loan_status')
X = pd.get_dummies(df.drop(columns='loan_status'))

# Create our target
targ = df['loan_status'].copy()
y = targ

In [34]:
X.describe()

loan_amnt      int_rate   installment    annual_inc           dti  \
count  68817.000000  68817.000000  68817.000000  6.881700e+04  68817.000000   
mean   16677.594562      0.127718    480.652863  8.821371e+04     21.778153   
std    10277.348590      0.048130    288.062432  1.155800e+05     20.199244   
min     1000.000000      0.060000     30.890000  4.000000e+01      0.000000   
25%     9000.000000      0.088100    265.730000  5.000000e+04     13.890000   
50%    15000.000000      0.118000    404.560000  7.300000e+04     19.760000   
75%    24000.000000      0.155700    648.100000  1.040000e+05     26.660000   
max    40000.000000      0.308400   1676.230000  8.797500e+06    999.000000   

        delinq_2yrs  inq_last_6mths      open_acc       pub_rec  \
count  68817.000000    68817.000000  68817.000000  68817.000000   
mean       0.217766        0.497697     12.587340      0.126030   
std        0.718367        0.758122      6.022869      0.336797   
min        0.000000        0.000000      2.000000      0.000000   
25%        0.000000        0.000000      8.000000      0.000000   
50%        0.000000        0.000000     11.000000      0.000000   
75%        0.000000        1.000000     16.000000      0.000000   
max       18.000000        5.000000     72.000000      4.000000   

           revol_bal     total_acc     out_prncp  out_prncp_inv   total_pymnt  \
count   68817.000000  68817.000000  68817.000000   68817.000000  68817.000000   
mean    17604.142828     24.666347  16018.845833   16015.919519    970.639473   
std     21835.880400     12.318917  10002.695495   10004.985940   1229.563301   
min         0.000000      3.000000      0.000000       0.000000      0.000000   
25%      6293.000000     16.000000   8578.800000    8576.500000    403.760000   
50%     12068.000000     23.000000  14067.750000   14061.030000    710.790000   
75%     21735.000000     31.000000  22860.700000   22860.490000   1214.600000   
max    587191.000000    164.000000  40000.000000   40000.000000  41418.981165   

       total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
count     68817.000000     68817.000000   68817.000000        68817.000000   
mean        970.431151       658.748750     311.855687            0.035072   
std        1229.569864      1126.237377     279.786954            1.086720   
min           0.000000         0.000000       0.000000            0.000000   
25%         403.550000       250.370000     115.440000            0.000000   
50%         710.790000       449.890000     225.250000            0.000000   
75%        1214.590000       762.170000     415.710000            0.000000   
max       41418.980000     40000.000000    2735.400000           75.240000   

       recoveries  collection_recovery_fee  last_pymnt_amnt  \
count     68817.0                  68817.0     68817.000000   
mean          0.0                      0.0       517.202218   
std           0.0                      0.0       816.021053   
min           0.0                      0.0         0.000000   
25%           0.0                      0.0       267.110000   
50%           0.0                      0.0       408.520000   
75%           0.0                      0.0       652.580000   
max           0.0                      0.0     40500.340000   

       collections_12_mths_ex_med  policy_code  acc_now_delinq   tot_coll_amt  \
count                68817.000000      68817.0         68817.0   68817.000000   
mean                     0.016188          1.0             0.0     188.664298   
std                      0.134774          0.0             0.0    1808.457195   
min                      0.000000          1.0             0.0       0.000000   
25%                      0.000000          1.0             0.0       0.000000   
50%                      0.000000          1.0             0.0       0.000000   
75%                      0.000000          1.0             0.0       0.000000   
max                      4.000000          1.

In [35]:
# Check the balance of our target values
y.value_counts()

low_risk     68470
high_risk      347
Name: loan_status, dtype: int64

In [36]:
# Split the X and y into X_train, X_test, y_train, y_test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [37]:
# Create the StandardScaler instance
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [38]:
# Fit the Standard Scaler with the training data
# When fitting scaling functions, only train on the training dataset
X_scaler = scaler.fit(X_train)

In [39]:
# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Display the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier only, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

### Balanced Random Forest Classifier

In [45]:
# Resample the training data with the BalancedRandomForestClassifier
from imblearn.ensemble import BalancedRandomForestClassifier, EasyEnsembleClassifier
model = BalancedRandomForestClassifier(n_estimators=100, random_state=1)

#fit model
model.fit(X_train_scaled, y_train)

#predict
y_predict = model.predict(X_test_scaled)

In [46]:
# Calculated the balanced accuracy score
print(f'BAS: {balanced_accuracy_score(y_test,y_predict)}')

BAS: 0.7887512850910909


In [47]:
# Display the confusion matrix
confusion_matrix(y_test, y_predict)

array([[   71,    30],
       [ 2146, 14958]])

In [48]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_predict))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.03      0.70      0.87      0.06      0.78      0.60       101
   low_risk       1.00      0.87      0.70      0.93      0.78      0.63     17104

avg / total       0.99      0.87      0.70      0.93      0.78      0.63     17205



In [50]:
# List the features sorted in descending order by feature importance
# sorted(zip(rf_model.feature_importances_, X.columns), reverse=True)
sorted(zip(model.feature_importances_, X.columns), reverse=True)[:10]

[(0.07876809003486353, 'total_rec_prncp'),
 (0.05883806887524815, 'total_pymnt'),
 (0.05625613759225244, 'total_pymnt_inv'),
 (0.05355513093134745, 'total_rec_int'),
 (0.0500331813446525, 'last_pymnt_amnt'),
 (0.02966959508700077, 'int_rate'),
 (0.021129125328012987, 'issue_d_Jan-2019'),
 (0.01980242888931366, 'installment'),
 (0.01747062730041245, 'dti'),
 (0.016858293184471483, 'out_prncp_inv')]

### Easy Ensemble Classifier

In [56]:
# Train the Classifier
easy_model = EasyEnsembleClassifier(n_estimators=100, random_state=1)

# Fit the model
easy_model.fit(X_train_scaled, y_train)

# Make the predictions
eas_y_pred = easy_model.predict(X_test_scaled)

In [57]:
# Calculated the balanced accuracy score
print(f'Easy E BAS: {balanced_accuracy_score(y_test, eas_y_pred)}')

Easy E BAS: 0.931601605553446


In [58]:
# Display the confusion matrix
confusion_matrix(y_test, eas_y_pred)

array([[   93,     8],
       [  985, 16119]])

In [59]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, eas_y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.09      0.92      0.94      0.16      0.93      0.87       101
   low_risk       1.00      0.94      0.92      0.97      0.93      0.87     17104

avg / total       0.99      0.94      0.92      0.97      0.93      0.87     17205



### Final Questions

1. Which model had the best balanced accuracy score?

    Easy Ensemble had the best balanced accuracy score.

2. Which model had the best recall score?

    The Easy Ensemble model had a high risk recall of .92 and low risk recall score of .94, which is much better than the Random Forrest. 

3. Which model had the best geometric mean score?

    Easy Enemble again was much better than the Random Forrest GM score. 

4. What are the top three features?

    Total received principle, total payment, and total payment inv, which probably means total payments invoiced.